### PR TITLE
IRIS-6405 | Use navigator.sendBeacon to mark notification as read

### PR DESCRIPTION
**Pull request by rogatty**
https://wikia-inc.atlassian.net/browse/IRIS-6405
https://developer.mozilla.org/en-US/docs/Web/API/Navigator/sendBeacon

@Wikia/iris

### CODE REVIEW

File: extensions/wikia/DesignSystem/scripts/DesignSystemGlobalNavigationOnSiteNotifications.controller.js [hunk 1]
```diff
@@ -114,22 +114,60 @@ define('ext.wikia.design-system.on-site-notifications.controller', [
...
+						contentType: 'application/json; charset=UTF-8',
+						// Keep it low as it's blocking user from navigating to the notification target
+						timeout: 500,
+						url: markAsReadUrl,
+						xhrFields: {
+							withCredentials: true
+						}
+					}).complete(function () {
+						window.location.href = notificationDetails.uri;
+					}.bind(this));
```
Comment (xkxd) [inline extensions/wikia/DesignSystem/scripts/DesignSystemGlobalNavigationOnSiteNotifications.controller.js]: binding is not necessary here

File: extensions/wikia/DesignSystem/scripts/DesignSystemGlobalNavigationOnSiteNotifications.controller.js [hunk 2]
```diff
@@ -114,22 +114,60 @@ define('ext.wikia.design-system.on-site-notifications.controller', [
...
 
+			markAsReadOnPageUnload: function (notificationDetails, event) {
+				if (!notificationDetails.isUnread) {
+					return;
+				}
+
+				var markAsReadUrl = this.getBaseUrl() + '/notifications/mark-as-read/by-uri';
+				var data = JSON.stringify([notificationDetails.uri]);
+
+				if (window.navigator.sendBeacon) {
```
Comment (xkxd) [inline extensions/wikia/DesignSystem/scripts/DesignSystemGlobalNavigationOnSiteNotifications.controller.js:145]: I'm wondering - how about using a polyfill here?

**Review by xkxd — APPROVED** [review]

File: extensions/wikia/DesignSystem/scripts/DesignSystemGlobalNavigationOnSiteNotifications.controller.js (hunk 2 above)
Comment (KockaAdmiralac) [reply extensions/wikia/DesignSystem/scripts/DesignSystemGlobalNavigationOnSiteNotifications.controller.js:145]: Wouldn't `navigator.sendBeacon` polyfills just be using synchronous `XMLHttpRequest`s (which are [deprecated](https://xhr.spec.whatwg.org/#the-open()-method))? Most of the polyfills for it I've seen are doing that...

File: extensions/wikia/DesignSystem/scripts/DesignSystemGlobalNavigationOnSiteNotifications.controller.js [hunk 3]
```diff
@@ -114,22 +114,72 @@ define('ext.wikia.design-system.on-site-notifications.controller', [
...
+					try {
+						var blob = new Blob([data], {
+							type: 'application/json'
+						});
+
+						if (window.navigator.sendBeacon(markAsReadUrl, blob) === false) {
```
Comment (vforge) [inline extensions/wikia/DesignSystem/scripts/DesignSystemGlobalNavigationOnSiteNotifications.controller.js]: If you'd check here for true (and return if it is true), you could use one `this.sendBeaconFallback` at the end

**Review by vforge — APPROVED** [review]
LGTM, one minor suggestion

File: extensions/wikia/DesignSystem/scripts/DesignSystemGlobalNavigationOnSiteNotifications.controller.js [hunk 4]
```diff
@@ -148,17 +148,17 @@ define('ext.wikia.design-system.on-site-notifications.controller', [
...
 
-						if (window.navigator.sendBeacon(markAsReadUrl, blob) === false) {
-							this.sendBeaconFallback(notificationDetails, data, markAsReadUrl);
+						if (window.navigator.sendBeacon(markAsReadUrl, blob) === true) {
+							return;
 						}
 					} catch (exception) {
 						log(exception, log.levels.warning, common.logTag);
 						// See http://crbug.com/490015#c99
 						this.sendBeaconFallback(notificationDetails, data, markAsReadUrl);
```
Comment (vforge) [inline extensions/wikia/DesignSystem/scripts/DesignSystemGlobalNavigationOnSiteNotifications.controller.js]: This is obsolete

**Pull request merged**